File: render/gpu-lock.mjs
// gpu-lock.mjs — the Node side of the machine-wide, FENCED GPU lease.
//
// THIS FILE NO LONGER ACQUIRES. internal/gpulease (Go) is the ONE implementation of
// acquisition, staleness, fencing and the epoch counter; a render INHERITS a lease that
// the Go caller already holds (GPU_LEASE_DIR / GPU_LEASE_EPOCH / GPU_LEASE_CLASS).
//
// WHY THE DUPLICATE WAS DELETED RATHER THAN FIXED. Two languages independently
// implementing one concurrency rule produced a new divergence in every review round:
// different atomic tokens (both sides ended up holding the lease), different liveness
// rules (EPERM meant "alive" here and "dead" there), a non-atomic epoch write that a
// measurement showed restarting the fence at 1 in 24.5% of concurrent reads, and one
// side deleting the other's in-progress claim. Each was fixed individually and the next
// round found the next one, because the defect was never a bug — it was the duplication.
// There is now exactly one rule, in one language, and this file is a consumer of it.
//
// What remains here is everything that is genuinely Node's job:
//   - honour an inherited lease and FENCE against it before anything irreversible,
//   - elect ONE unloader per lease so a batch costs one teardown, not N,
//   - DRAIN llama-swap before unloading, because the unload route does not,
//   - the ComfyUI lifecycle and the guarded teardown.
//
// No npm dependencies.
import { writeFileSync, readFileSync, unlinkSync } from "node:fs";
import { join } from "node:path";
import { ensureComfy as defaultEnsureComfy } from "./comfy-lifecycle.mjs";

function metaPath(lockPath) {
  return join(lockPath, "meta.json");
}

// readLease parses the shared lease record written by internal/gpulease. Node only ever
// READS it — the schema is owned on the Go side.
export function readLease(lockPath) {
  try {
    return JSON.parse(readFileSync(metaPath(lockPath), "utf8"));
  } catch { return null; }
}

// inheritedLease: the Go holder threads its lease down. Absent env means no lease, which
// is now an ERROR for a GPU job rather than a cue to acquire one (see withGpuSlot).
export function inheritedLease(env = process.env) {
  const dir = (env.GPU_LEASE_DIR || "").trim();
  const epoch = (env.GPU_LEASE_EPOCH || "").trim();
  if (!dir || !epoch) return null;
  return { dir, epoch: Number(epoch), class: (env.GPU_LEASE_CLASS || "").trim() || undefined };
}

// checkInheritedLease is the FENCE. Before an irreversible action, confirm the epoch we
// were handed is still current. A closing laptop lid is not a crash — the process
// survives and resumes, and without this it would resume and unload models on top of
// whoever holds the card now, which is the original incident replayed by a lid.
export function checkInheritedLease(lease, read = readLease) {
  if (!lease) return true;
  const meta = read(lease.dir);
  if (!meta) return false;
  return Number(meta.epoch) === Number(lease.epoch);
}

// claimLeaseUnload: elect the ONE job that unloads for this lease. Exclusive creation of
// a per-epoch marker makes the winner unambiguous even when several jobs start under one
// lease at once. Markers are cleaned up by the Go release path.
//
// This is what makes the hoist real. Skipping the unload under an inherited lease while
// nothing performed it left a leased render running with every model resident.
export function claimLeaseUnload(lease) {
  if (!lease || !lease.dir || !Number.isFinite(lease.epoch)) return true;
  try {
    writeFileSync(join(lease.dir, `unloaded.${lease.epoch}`), String(process.pid), { flag: "wx" });
    return true;
  } catch (e) {
    if (e.code === "EEXIST") return false; // someone already unloaded for this lease
    return true; // cannot tell => unload. Correctness over saving one teardown.
  }
}

// releaseLeaseUnloadMarker lets a standalone run clean up after itself. The Go release
// path sweeps these too; this is belt and braces for a job that owns its own marker.
export function releaseLeaseUnloadMarker(lease) {
  if (!lease) return;
  try { unlinkSync(join(lease.dir, `unloaded.${lease.epoch}`)); } catch {}
}

// MEMORY_STACK: the always-loaded, CPU-only mem0 models (they hold ZERO GPU VRAM).
// freeLlamaSwap must NEVER unload these — the unload-ALL route did, needlessly tearing
// down the load-bearing memory stack on every gen job for no VRAM benefit.
//
// SOURCED FROM CONFIG/ENV, not a buried const: the Go harness threads the config's
// MemoryStack as MEMORY_STACK, so a renamed/added 3rd CPU member is honored instead
// of silently unloaded. The literal below is the fallback for a direct CLI run.
const DEFAULT_MEMORY_STACK = ["embeddinggemma", "bge-reranker-v2-m3"];
export function memoryStack(env = process.env.MEMORY_STACK) {
  if (env && env.trim()) {
    return new Set(env.split(",").map((s) => s.trim()).filter(Boolean));
  }
  return new Set(DEFAULT_MEMORY_STACK);
}

// withTimeout: every network call here needs its own deadline. Node's fetch has NO
// default request timeout, so a socket that accepts and then stalls would hang the
// drain — and with it the render — indefinitely, regardless of any timeoutMs we track.
function withTimeout(ms) {
  return AbortSignal.timeout(ms);
}

// quiesceLlamaSwap: wait for in-flight work on `ids` to finish before unloading them.
//
// MEASURED, not defensive: on llama-swap v242 an unload issued during a generation
// returned in 1,265ms without draining and the in-flight request died at 4,107ms with
// 502 Bad Gateway. The unload route does not honour in-flight work, so a caller that
// wants to avoid killing someone's request must drain first.
//
// Signal: llama-server's own /slots via llama-swap's /upstream/<id>/slots, where
// is_processing is true exactly while a slot is generating (verified against a
// 23s/1500-token generation).
//
// FAIL-SAFE, NOT FAIL-OPEN-SILENT. A 404 is only "idle" when the model is genuinely
// NOT LOADED — it is also what a loaded upstream without a /slots route returns (any
// non-llama.cpp backend on :11436; whisper is one). Accepting that as idle reported a
// verified drain while in-flight work was killed, which is the exact 502 this exists to
// prevent. So a 404 is cross-checked against /running, and anything we cannot observe
// is named in `unknown` rather than assumed quiet.
export async function quiesceLlamaSwap(ids, {
  api = process.env.LLAMA_SWAP_API || "http://localhost:11436",
  timeoutMs = 60_000, pollMs = 500, graceMs = 1_500,
  fetchImpl = fetch, nowFn = Date.now, requestTimeoutMs = 5_000,
} = {}) {
  const deadline = nowFn() + timeoutMs;
  const unknown = new Set();
  const started = nowFn();

  // Which models does the server consider loaded? Used to disambiguate a 404.
  const loaded = async () => {
    try {
      const r = await fetchImpl(`${api}/running`, { signal: withTimeout(requestTimeoutMs) });
      if (!r.ok) return null;
      const j = await r.json();
      return new Set((j.running || []).map((m) => m.model).filter(Boolean));
    } catch { return null; }
  };

  // /running GATES every probe. On llama-swap v208 (live-found on <node-c>),
  // `/upstream/<id>/slots` is `Any /upstream/*` -> proxyToUpstream: requesting it
  // for a model that is NOT loaded swaps the model IN — the drain was loading ~3GB
  // into VRAM immediately before the render it existed to protect. So:
  //   - id not in /running  => nothing to drain, and NO probe is ever sent;
  //   - /running unreadable => we are blind, and hands-off beats a probe that may
  //     LOAD a model: every id is named unknown, no /upstream request fires.
  const busy = async (id, loadedSet) => {
    if (!loadedSet) { unknown.add(id); return false; } // blind: never probe
    if (!loadedSet.has(id)) return false;              // not loaded: nothing in flight
    try {
      const r = await fetchImpl(`${api}/upstream/${id}/slots`, { signal: withTimeout(requestTimeoutMs) });
      if (r.status === 404) {
        // Loaded but no /slots route (whisper, any non-llama.cpp backend): we are
        // blind to it, and must say so instead of claiming a drain.
        unknown.add(id);
        return false;
      }
      if (!r.ok) { unknown.add(id); return false; }
      const j = await r.json();
      const slots = Array.isArray(j) ? j : [j];
      return slots.some((s) => s && s.is_processing === true);
    } catch { unknown.add(id); return false; }
  };

  // An unobservable tier reads as "not busy", which would otherwise end the drain on
  // the first transient blip. Retry a BOUNDED number of extra rounds so a momentary
  // 502 (a model swapping, say) resolves, while a permanently unobservable tier — one
  // with no /slots route at all — still costs only a few polls instead of the whole
  // timeout before every render.
  const unknownRetries = 2;
  let unknownRounds = 0;
  for (;;) {
    unknown.clear(); // judge each round on its own evidence, not a stale transient
    const loadedSet = await loaded();
    const flags = await Promise.all(ids.map((id) => busy(id, loadedSet)));
    if (!flags.some(Boolean)) {
      if (unknown.size === 0) break;               // verified idle
      if (++unknownRounds > unknownRetries) break; // accept: genuinely unobservable
    } else {
      unknownRounds = 0;
    }
    if (nowFn() >= deadline) {
      return { drained: false, waitedMs: nowFn() - started, unknown: [...unknown] };
    }
    await new Promise((r) => setTimeout(r, pollMs));
  }
  if (unknown.size > 0) {
    // We could not observe some tiers. Give in-flight work a brief grace rather than
    // claiming a drain we did not verify.
    await new Promise((r) => setTimeout(r, graceMs));
  }
  return { drained: unknown.size === 0, waitedMs: nowFn() - started, unknown: [...unknown] };
}

// freeLlamaSwap: free the GPU-resident llama-swap models so their VRAM goes to a gen
// job, while leaving the CPU memory stack warm. DRAINS FIRST (see quiesceLlamaSwap).
//
// Called ONCE PER LEASE, not per job — withGpuSlot enforces that via claimLeaseUnload.
// Errors are reported through `log` rather than swallowed: silently unloading nothing
// leaves the render to OOM against a full card.
export async function freeLlamaSwap(api = process.env.LLAMA_SWAP_API || "http://localhost:11436", opts = {}) {
  const {
    quiesce = quiesceLlamaSwap, drainTimeoutMs = 60_000, requestTimeoutMs = 10_000,
    log = (m) => console.error(m),
  } = opts;
  const keep = memoryStack();
  let ids = [];
  try {
    const r = await fetch(api + "/v1/models", { signal: withTimeout(requestTimeoutMs) });
    if (!r.ok) { log(`freeLlamaSwap: /v1/models returned ${r.status}; NOT unloading (the render keeps a shared card)`); return; }
    const j = await r.json();
    ids = (j.data || []).map((m) => m.id).filter((id) => id && !keep.has(id));
  } catch (e) {
    log(`freeLlamaSwap: could not list models (${e && e.message}); NOT unloading (the render keeps a shared card)`);
    return;
  }
  if (ids.length === 0) return;

  // /running decides what actually holds VRAM. Unloading a model that is not
  // loaded is not merely pointless: on llama-swap v208 the per-model unload
  // route does not even exist, so each such request was a full requestTimeoutMs
  // of dead wait per configured model, on every render. And if /running cannot
  // be read, unloading blind is how a shared card gets a load-bearing tier torn
  // down — hands off, loudly, like the /v1/models failure above.
  let running;
  try {
    const r = await fetch(api + "/running", { signal: withTimeout(requestTimeoutMs) });
    if (!r.ok) { log(`freeLlamaSwap: /running returned ${r.status}; NOT unloading (cannot see what is loaded)`); return; }
    const j = await r.json();
    running = new Set((j.running || []).map((m) => m.model).filter(Boolean));
  } catch (e) {
    log(`freeLlamaSwap: could not read /running (${e && e.message}); NOT unloading (cannot see what is loaded)`);
    return;
  }
  const loaded = ids.filter((id) => running.has(id));
  if (loaded.length === 0) return; // nothing of ours is holding VRAM

  try {
    const res = await quiesce(loaded, { api, timeoutMs: drainTimeoutMs });
    if (!res.drained) {
      // Loud, not silent: proceeding here may kill someone's in-flight request. The
      // alternative — blocking the render forever behind a stuck tier — is worse, so
      // we proceed and say so.
      log(`freeLlamaSwap: proceeding without a verified drain after ${res.waitedMs}ms` +
          (res.unknown.length ? ` (could not read /slots for: ${res.unknown.join(",")})` : ""));
    }
  } catch (e) {
    log(`freeLlamaSwap: drain failed (${e && e.message}); proceeding to unload`);
  }

  // Per-model unload first (llama-swap >= v24x). Any failure falls back ONCE to
  // v208's only unload route — GET /unload, which unloads EVERYTHING — and that
  // fallback is gated on the memory stack not being resident: tearing down the
  // always-on CPU tier to free VRAM it does not hold would trade a render for
  // the memory stack (invariant 1).
  const failures = [];
  await Promise.all(loaded.map(async (id) => {
    try {
      const r = await fetch(api + "/api/models/unload/" + id, { method: "POST", signal: withTimeout(requestTimeoutMs) });
      if (!r.ok) failures.push(id);
    } catch (e) {
      log(`freeLlamaSwap: unload ${id} failed: ${e && e.message}`);
      failures.push(id);
    }
  }));
  if (failures.length === 0) return;

  const keepResident = [...running].filter((id) => keep.has(id));
  if (keepResident.length > 0) {
    log(`freeLlamaSwap: per-model unload unavailable for ${failures.join(",")} and the memory stack (${keepResident.join(",")}) is resident — cannot unload-all; the render runs against whatever VRAM remains`);
    return;
  }
  try {
    const r = await fetch(api + "/unload", { signal: withTimeout(requestTimeoutMs) });
    if (r.ok) log(`freeLlamaSwap: per-model unload unavailable (${failures.join(",")}); GET /unload (unload-all, llama-swap v208 route) succeeded`);
    else log(`freeLlamaSwap: unload-all fallback returned ${r.status}; the render runs against whatever VRAM remains`);
  } catch (e) {
    log(`freeLlamaSwap: unload-all fallback failed: ${e && e.message}`);
  }
}

// freeComfy: tell ComfyUI to drop loaded models + free VRAM after a job (zero-warm).
export async function freeComfy(api = process.env.COMFY_API || "http://127.0.0.1:8188") {
  try {
    await fetch(api + "/free", {
      method: "POST", headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ unload_models: true, free_memory: true }),
      signal: withTimeout(10_000),
    });
  } catch {}
}

// withGpuSlot centralizes the single-slot GPU lifecycle every gen runner shares:
//   1. REQUIRE a lease (inherited from the Go caller) unless noLock — this file no
//      longer acquires, so a GPU job without one is a wiring error, not a cue to grab
//      the card,
//   2. FENCE against it, then unload llama-swap ONCE PER LEASE and only for a `media`
//      lease — a `text` lease is a benchmark's reservation and unloading under it
//      destroys exactly the run it was taken to protect,
//   3. optionally ensureComfy(); warm:true is the BATCH-SESSION mode,
//   4. await fn(),
//   5. run ONE guarded teardown: freeComfy() + kill a ComfyUI we spawned.
// Deps (freeLlamaSwap/ensureComfy/freeComfy/checkLease/claimUnload) are injectable for
// tests only.
export async function withGpuSlot(opts, fn) {
  const {
    noLock = false,
    keepComfy = false,
    comfyManaged = true,
    warm = false,
    reserveVram,
    freeLlamaSwap: freeLS = freeLlamaSwap,
    ensureComfy = defaultEnsureComfy,
    freeComfy: freeCfy = freeComfy,
    lease = inheritedLease(),
    claimUnload = claimLeaseUnload,
    checkLease = checkInheritedLease,
  } = opts || {};

  // No lease, and not explicitly opted out => refuse. Acquiring here is exactly the
  // duplicate implementation that was deleted; silently rendering unarbitrated is the
  // behaviour that tore the text tier down in the first place.
  if (!noLock && !lease) {
    throw new Error(
      "GPU lease missing: this runner no longer acquires the GPU itself. " +
      "Run it under the harness (which takes the lease and threads GPU_LEASE_DIR/EPOCH/CLASS), " +
      "or for a standalone run wrap it: `local-offload gpu reserve --class media -- node <script> ...`. " +
      "Use --no-lock only when you know nothing else can touch the GPU.");
  }

  let comfyChild = null;
  let cleaning = false;
  const cleanup = async () => {
    if (cleaning) return; cleaning = true;
    if (comfyManaged) { try { await freeCfy(); } catch {} }
    if (comfyChild && !keepComfy) { try { comfyChild.kill(); } catch {} }
  };
  const onSig = async () => { await cleanup(); process.exit(130); };
  for (const sig of ["SIGINT", "SIGTERM", "SIGBREAK"]) process.on(sig, onSig);
  try {
    // THE FENCE COMES FIRST, AND IT IS UNCONDITIONAL. It used to sit inside the
    // unload-election branch, so it was skipped for every job that lost the election
    // (jobs 2..N of a batch) and for every `text` lease — those jobs went straight to
    // submitting a graph while fenced out, i.e. rendering on somebody else's card.
    // Submitting a graph is irreversible GPU work, so it needs the same guard the
    // unload does.
    if (lease && !checkLease(lease)) {
      throw new Error(
        `GPU lease epoch ${lease.epoch} is no longer current — this process was fenced out ` +
        `(the card was handed to another holder while we were suspended). Refusing to touch the GPU.`);
    }
    // Then the class gate, then the once-per-lease election.
    const mayUnload = !lease || lease.class !== "text";
    if (mayUnload && (!lease || claimUnload(lease))) {
      await freeLS();
    }
    if (comfyManaged) {
      comfyChild = await ensureComfy({
        ...(reserveVram != null ? { reserveVram } : {}),
        ...(warm ? { warm: true } : {}),
      });
    }
    return await fn({ comfyChild, lease });
  } finally {
    await cleanup();
    for (const sig of ["SIGINT", "SIGTERM", "SIGBREAK"]) process.removeListener(sig, onSig);
  }
}

File: internal/gpulock/gpulock.go
// Package gpulock is the READ-ONLY Go view of the GPU lease that internal/gpulease
// owns and that every GPU consumer takes. The GPU is shared between llama-swap (the
// VLM/text tiers) and ComfyUI/Chatterbox (the generation jobs).
//
// LO-1 evidence: while a generation job held the lease, llama-swap could not (re)load
// the vision model, so EVERY vision call 5xx'd and DEFERRED — 295 of the 337 all-time
// defers landed inside one such hour. A defer is not a cloud API call: the harness
// never calls a cloud model (see main.go), it returns a structured defer and the work
// falls back to Opus, the calling session. The cost is that the harness stops doing
// the one thing it exists for, silently. This package lets the pipeline SEE the lease
// before burning a doomed HTTP call: resolve the same path every other consumer uses,
// report held/not-held plus holder age, and wait (bounded) for the slot to free.
//
// Invariants:
//   - Read-only: this package NEVER creates, reclaims, or removes the lease —
//     acquisition and stale-reclaim belong to internal/gpulease.
//   - It resolves the SAME path and parses the SAME record as every other
//     participant, and shares one definition of process liveness with them. Each of
//     those was independently wrong once, and each time the symptom was identical:
//     this package reported a live holder's lease as FREE.
package gpulock

import (
	"context"
	"os"
	"path/filepath"
	"time"

	"github.com/dmmdea/offload-harness/internal/gpulease"
)

// DefaultTTL mirrors DEFAULT_TTL_MS in render/gpu-lock.mjs (1h — a real video
// generation can take many minutes; the TTL is only the fallback when the
// holder pid is unknown, and the backstop against pid recycling).
const DefaultTTL = time.Hour

// Path resolves the shared GPU lease directory exactly the way every other consumer
// does: explicit override first (the gpu_lock_path config field — the pipeline also
// threads it to the runners as the GPU_LOCK env so both sides always contend on ONE
// path), then the GPU_LOCK env, then <state-root>/gpu/lease.
//
// THE DEFAULT MOVED, AND THIS MUST MOVE WITH IT. The lease used to live under the OS
// temp dir. When it became machine-wide (%ProgramData%) this function was left behind,
// so the vision gate watched a path nothing writes and WaitFree answered "free" every
// time — silently reinstating the LO-1 regression it exists to prevent (a vision call
// fired into a busy GPU 5xx's and defers, so the harness stops offloading at exactly
// the moment offloading matters). Resolution is delegated rather than duplicated so
// the two can never drift apart again.
func Path(override, stateDir string) string {
	// ONE resolver for every consumer. Re-deriving the location here is what split the
	// lease in two: this package honoured the override while the acquirer ignored it,
	// so the gate watched a directory nothing wrote.
	dir, err := gpulease.LeaseDir(override, stateDir)
	if err != nil {
		// Read-only inspection must never fail the caller, but it must not silently
		// pretend either: an unresolvable location degrades to a path that provably
		// does not exist, so Inspect reports NOT held and the caller proceeds — the
		// same outcome as today, without inventing a plausible-looking tmpdir path
		// that could collide with a real lock.
		return filepath.Join(os.TempDir(), "local-offload-gpu-UNRESOLVED")
	}
	return dir
}

// Info is one point-in-time inspection of the lease.
type Info struct {
	// Held is true when a claim exists AND the shared staleness rule says it is live.
	Held bool
	// Age is how long the current holder has held the card, from the acquirer's own
	// stamp. NOT time-since-last-heartbeat, which reads as a few seconds for any
	// long-running holder and made the defer message report "(0s)" for a job that had
	// owned the card for an hour.
	Age time.Duration
	// PID is the recorded holder pid (0 when unknown / not held).
	PID int
	// Class is the holder's lease class (media/text), so a caller can say WHICH kind of
	// work holds the card instead of assuming it is a generation job.
	Class string
	// ExpiresAt is the holder's DECLARED window end (zero when it declared none).
	// WaitFree uses it to stop polling a reservation that provably cannot free
	// inside the caller's wait: before 0.113.27 every vision call burned its
	// full vision_gpu_wait_sec (90 s by default) against a multi-hour hold, so
	// a long reservation cost 90 s of dead wait PER CALL for its whole life.
	ExpiresAt time.Time
}

// Inspect reports whether the lock at lockPath is currently held, using the
// same staleness rule as gpu-lock.mjs with the default 1h TTL. A stale lock
// (dead holder / no meta / over-TTL) reports NOT held — reclaiming it is the
// runners' job, never ours.
func Inspect(lockPath string) Info { return inspectAt(lockPath, DefaultTTL, time.Now()) }

// inspectAt is Inspect with an injectable TTL and clock, retained for the existing
// unit tests. The TTL argument is now ADVISORY ONLY: staleness is decided by
// gpulease.Reclaimable, not here.
//
// THE THIRD RULE IS GONE. This package used to carry its own staleness logic —
// (pid dead) OR (meta.json mtime older than 1h) — which was a THIRD implementation
// alongside Go's and Node's, and it disagreed with both: gpulease reclaims on
// (provably gone) OR (recycled pid) OR (heartbeat stale AND declared window expired).
// A holder that heartbeats for over an hour is live to gpulease and stale here, so this
// gate would report a busy card as free. That is the same failure mode as the path split
// (C5) and the schema split, one layer down. Delegating means there is one rule.
func inspectAt(lockPath string, _ time.Duration, _ time.Time) Info {
	// Delegate WHOLESALE, not just the rule. Reconstructing the judgement here from the
	// record — even using the shared Reclaimable — was still a second reader, and it
	// broke the moment the heartbeat moved into a per-epoch file: a record-only view
	// sees a frozen RenewedAtMs and calls a live, renewing holder stale as soon as its
	// declared window lapses. One inspection path, one answer.
	i := gpulease.InspectDir(lockPath)
	return Info{Held: i.Held, Age: i.Age, PID: i.PID, Class: string(i.Class), ExpiresAt: i.ExpiresAt}
}

// WaitFree polls the lock every poll (min bound 1ms; the pipeline passes 2s)
// until it is free or wait has elapsed (or ctx is done), returning the FINAL
// inspection: Held=false means the slot freed (proceed); Held=true means the
// caller should defer, with Age available for the defer reason.
func WaitFree(ctx context.Context, lockPath string, wait, poll time.Duration) Info {
	if poll <= 0 {
		poll = time.Millisecond
	}
	deadline := time.Now().Add(wait)
	for {
		info := Inspect(lockPath)
		if !info.Held {
			return info
		}
		// NEVER WAIT FOR SOMETHING THAT CANNOT HAPPEN (0.113.27) — the same
		// rule, and the same reasoning, as gpulease.Acquire's short-circuit.
		// A TEXT reservation carries an operator's declared duration
		// (`gpu reserve --for 8h`), so if it outlasts our whole window there
		// is nothing to wait for: the vision lane used to burn its full
		// vision_gpu_wait_sec on every call for the entire hold.
		//
		// A MEDIA lease is deliberately excluded: its expiry is a timeout
		// CEILING, not a promise — a video job declares 25 minutes and
		// routinely finishes in three — so short-circuiting on it would defer
		// calls that were about to be served. A holder with no declared
		// window is polled exactly as before.
		if info.Class == string(gpulease.ClassText) && !info.ExpiresAt.IsZero() && info.ExpiresAt.After(deadline) {
			return info
		}
		remain := time.Until(deadline)
		if remain <= 0 || ctx.Err() != nil {
			return info
		}
		if poll < remain {
			remain = poll
		}
		select {
		case <-ctx.Done():
			return info
		case <-time.After(remain):
		}
	}
}
